lib/spack/spack/binary_distribution.py

```diff
@@ -252,10 +259,19 @@ def _associate_built_specs_with_mirror(self, cache_key, url_and_version: MirrorU
             db = BuildCacheDatabase(tmpdir)
 
             try:
-                self._index_file_cache.init_entry(cache_key)
+                cache_exists = self._index_file_cache.init_entry(cache_key)
                 cache_path = self._index_file_cache.cache_path(cache_key)
-                with self._index_file_cache.read_transaction(cache_key):
-                    db._read_from_file(pathlib.Path(cache_path))
+                if not cache_exists:
+                    # recreate index if it is missing
+                    cache_entry = self._local_index_cache[cache_key]
+                    self._fetch_and_cache_index(url_and_version, cache_entry)
+                    cache_exists = self._index_file_cache.init_entry(cache_key)
+                if cache_exists:
+                    with self._index_file_cache.read_transaction(cache_key):
+                        db._read_from_file(cache_path)
+                else:
+                    # this is a logic bug, we should never hit this
+                    raise FileNotFoundError(f"cache index file missing: {cache_path}")
```

Comment:
```suggestion
                with self._index_file_cache.read_transaction(cache_key):
                    if not cache_exists:
                        # recreate index if it is missing
                        cache_entry = self._local_index_cache[cache_key]
                        self._fetch_and_cache_index(url_and_version, cache_entry)
                    db._read_from_file(cache_path)
```
I think there is still a race condition as written, but that this would eliminate it

Reply:
I think this would need  a write transaction no. Testing on #50857

Reply:
Yes, it would